crates/bevy_ecs/src/reflect.rs

```diff
@@ -8,9 +8,23 @@ use crate::{
 };
 use bevy_reflect::{impl_reflect_value, FromType, Reflect, ReflectDeserialize};
 
+/// A runtime type-reflectable component
+///
+/// Intended for use with [`bevy_reflect`],
+/// a [`ReflectComponent`] is a type-erased version of a component's data,
+/// can be transformed into  ['dyn Reflect'](Reflect) trait objects,
+/// which can be worked with generically
+/// and loaded from disk in a type-safe fashion.
+///
+/// [`ReflectComponent`] objects are created for a particular [`Component`] type (`C`) using the [`from_type`](FromType::from_type) method.
+/// That type `C` is implicitly stored in the function pointers held within the private fields of this type;
+/// it cannot be changed after creation.
+///
+/// Once a [`ReflectComponent`] object has been created, you can use that concrete struct
+/// to use the methods on this type, which always implicitly affect only the component type originally used to create this struct.
```

Comment:
This could really use an example. The interaction between `ReflectComponent` and any given `&dyn Reflect` is profoundly unintuitive.

Also, nit: "always implicitly affects only..." to me sounds like it's suggesting you can use these methods without worrying about types, because they handle the checking for you. In actual fact the situation is a bit more dire; some of these methods panic if the reference you give them is of the wrong type.